connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java

```diff
@@ -118,11 +140,73 @@ public void report(ProcessingContext context) {
                     originalMessage.key(), originalMessage.value(), originalMessage.headers());
         }
 
+        if (connConfig.isDlqContextHeadersEnabled()) {
+            populateContextHeaders(producerRecord, context);
+        }
+
         this.kafkaProducer.send(producerRecord, (metadata, exception) -> {
             if (exception != null) {
                 log.error("Could not produce message to dead letter queue. topic=" + dlqTopicName, exception);
                 errorHandlingMetrics.recordDeadLetterQueueProduceFailed();
             }
         });
     }
+
+    // Visible for testing
+    void populateContextHeaders(ProducerRecord<byte[], byte[]> producerRecord, ProcessingContext context) {
+        String topic = "";
+        int partition = -1;
+        long offset = -1;
+        if (context.consumerRecord() != null) {
+            topic = context.consumerRecord().topic();
+            partition = context.consumerRecord().partition();
+            offset = context.consumerRecord().offset();
+        }
+
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_TOPIC), topic);
```

Comment:
I don't think we want to compute the header name for every record. Why not simply define the header name constants to include the prefix?